### PR TITLE
all __ne__ methods removed

### LINKED ISSUE #23202: Dropping "all" `__ne__` methods?
Quite a few of the `__ne__` methods look like:
https://github.com/sympy/sympy/blob/0bc2587aef09ab150bd408b7cd9e08baf7b5fd9c/sympy/polys/polytools.py#L4251-L4252

In Python 3 one get an inverted `__eq__` for free, so no need to implement it if it just looks like that.

Are there any arguments against just deleting all methods that looks like that (and maybe some more that are effectively the inversion of the `__eq__` method)? (As long as the tests pass etc.)

See, e.g.
* https://stackoverflow.com/questions/4352244/should-ne-be-implemented-as-the-negation-of-eq
* https://bugs.python.org/issue4395

### PR DESCRIPTION
**Pull request by AsRaNi1**
# Removed all `__ne__` methods

Fixes: #23202 

#### Brief description of what is fixed or changed
All the undesirable ``__ne__`` methods have been removed as said

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by anutosh491** [comment]
@AsRaNi1 As pointed out in the error message above please add an appropriate release notes depicting your change !

**Comment by AsRaNi1** [comment]
> @AsRaNi1 As pointed out in the error message above please add appropriate release notes depicting your change !

@anutosh491 I have done the required changes, Thankyou!

**Comment by faze-geek** [comment]
@AsRaNi1 Firstly you can always run `python bin/test quality` on your terminal to see if code quality fails before committing anything .It can fail due to multiple reasons.  You will need to correct that since without it other tests would not run .

**Comment by faze-geek** [comment]
See https://github.com/sympy/sympy/wiki/Development-workflow#add-your-name-and-email-address-to-the-mailmap-file how to add yourself to the authors file.

**Comment by AsRaNi1** [comment]
@faze-geek is there any reason in particular that these tests are taking so long to run or do these take a long time in general?

**Comment by anutosh491** [comment]
> @faze-geek is there any reason in particular that these tests are taking so long to run or do these take a long time in general?

Nope there is some error from one of your commits . Mostly the test would auto cancel (usually after 3 hrs) and show that the latest commit wasn't correct ,but this has happened a couple of times on one of my prs too ( tests run infinitely even after 3 hrs ,mine had failed after a day)

**Comment by oscargus** [comment]
It is in general a bad idea to use the master branch for PRs. One problem is that you right now cannot easily rebase the PR and squash the commits. 

I'd like to encourage you to create a separate branch for this PR, but it is non-trivial if you do not want to do all edits again (but it it possible with some googling).

**Comment by anutosh491** [comment]
Yeah correct ,as pointed out above  working on a seperate branch is always better , if you continue working here , you might have to follow some steps to restore your master branch correctly with that of sympy once this goes in. You could follow this for restoring it https://github.com/sympy/sympy/pull/22872#issuecomment-1031365942

**Comment by oscargus** [comment]
I'm closing and reopening to trigger the tests again. This looks OK, but I guess only the tests can tell if some of the removed cases were actually needed.

File: .mailmap [hunk 1]
```diff
@@ -199,6 +199,8 @@ Archit Verma <architv07@gmail.com>
 Arif Ahmed <arif.ahmed.5.10.1995@gmail.com>
 Arighna Chakrabarty <arighna.chakrabarty100@gmail.com>
 Arihant Parsoya <parsoyaarihant@gmail.com>
+Arnav Zutshi <84843295+AsRaNi1@users.noreply.github.com>
+Arnav Zutshi <arnzut1324@gmail.com>
```
Comment (oscargus) [inline .mailmap:203]:
```suggestion
Arnav Zutshi <arnzut1324@gmail.com> Arnav Zutshi <84843295+AsRaNi1@users.noreply.github.com>
```
The current version will give two entries in the author list. The suggested change will map both emails to a single entry.

**Comment by oscargus** [comment]
Regarding release note, I would claim this is an internal change and nothing that a user would have to consider. So I will probably remove it (replace with `NO ENTRY`) if I merge soon.

**Comment by oscargus** [comment]
Ahh, it times out! That is why...

So then one would have to look a bit more careful at which test is timing out and try to figure out which `__ne__`-method that must be kept.

So far, I can probably say that all that looks like `return not self == other` can be removed. There may be a reason that the others are written in a different way (or there may not be). Also, if I understand correctly, sometimes one can still need the "trivial" above, in case one subclass a method which defines its own `__ne__`... But not sure here.

**Comment by oscarbenjamin** [comment]
The following test from `test_normalforms.py` hangs:
```python
from sympy import *
from sympy.matrices.normalforms import invariant_factors

x = Symbol('x')
m = Matrix([[Poly(x-1), Poly(1, x),Poly(-1,x)],
            [0, Poly(x), Poly(-1,x)],
            [Poly(0,x),Poly(-1,x),Poly(x)]])
invs = 1, x - 1, x**2 - 1
assert invariant_factors(m, domain=QQ[x]) == invs
```
The reason for this is:
```python
In [2]: p = QQ[x].from_sympy(0)

In [3]: p
Out[3]: 0

In [4]: type(p)
Out[4]: sympy.polys.rings.PolyElement

In [5]: p == 0
Out[5]: True

In [6]: p != 0
Out[6]: True
```
That causes this to be an infinite loop:
https://github.com/sympy/sympy/blob/94a6ee85d9d74c0d4fb49a18d9f576267b093b64/sympy/polys/matrices/normalforms.py#L129-L132

**Comment by oscarbenjamin** [comment]
As described in the original SO post:
https://stackoverflow.com/a/30676267/9450991
```
The __ne__ method follows automatically from __eq__ only if __ne__ isn't already defined in a superclass. So, if you're inheriting from a builtin, it's best to override both.
```
PolyElement is a subclass of dict so `__ne__` needs to be defined:
```python
In [13]: p = QQ[x].from_sympy(0)

In [14]: type(p).mro()
Out[14]: 
[sympy.polys.rings.PolyElement,
 sympy.polys.rings.PolyElement,
 sympy.polys.domains.domainelement.DomainElement,
 sympy.printing.defaults.Printable,
 sympy.core.sympify.CantSympify,
 dict,
 object]
```
All the other classes that are modified here should be checked. Also it should be checked that there are tests covering `!=` for all of these classes.

**Comment by oscargus** [comment]
> Also it should be checked that there are tests covering `!=` for all of these classes.

We should try to introduce code coverage again... Unfortunate that the code base is so big so that basically no provider can handle it...

**Comment by oscarbenjamin** [comment]
> We should try to introduce code coverage again.

It can be done with GitHub artefacts. We could use something like diff-cover and have it reported to the PR in the same way as the benchmarks:
https://diff-cover.readthedocs.io/en/latest/README.html
Full coverage reports could run on master and produce less frequent but larger artefacts.

**Comment by oscargus** [comment]
diff-cover would for sure be valuable.

However, if we just had some previous full coverage in a suitable format it would be possible to *know* which `__ne__` methods (and parts of) that are already covered. Currently a quite non-trivial exercise for most contributors.

**Comment by oscarbenjamin** [comment]
> However, if we just had some previous full coverage in a suitable format it would be possible to _know_ which `__ne__` methods (and parts of) that are already covered.

A simple possibility would be to add a GA workflow for the master branch that builds the html coverage report and pushes it to a repo like `sympy/sympy_coverage` under e.g. the `master` folder. That could then be hosted via GH-pages. We could extend that with a PR workflow that pushed to `prs/12345` and have comment with a link added to the PR so that you can see full coverage reports for a PR as well.

**Comment by AsRaNi1** [comment]
> The following test from `test_normalforms.py` hangs:
> 
> ```python
> from sympy import *
> from sympy.matrices.normalforms import invariant_factors
> 
> x = Symbol('x')
> m = Matrix([[Poly(x-1), Poly(1, x),Poly(-1,x)],
>             [0, Poly(x), Poly(-1,x)],
>             [Poly(0,x),Poly(-1,x),Poly(x)]])
> invs = 1, x - 1, x**2 - 1
> assert invariant_factors(m, domain=QQ[x]) == invs
> ```
> 
> The reason for this is:
> 
> ```python
> In [2]: p = QQ[x].from_sympy(0)
> 
> In [3]: p
> Out[3]: 0
> 
> In [4]: type(p)
> Out[4]: sympy.polys.rings.PolyElement
> 
> In [5]: p == 0
> Out[5]: True
> 
> In [6]: p != 0
> Out[6]: True
> ```
> 
> That causes this to be an infinite loop:
> 
> https://github.com/sympy/sympy/blob/94a6ee85d9d74c0d4fb49a18d9f576267b093b64/sympy/polys/matrices/normalforms.py#L129-L132

I will look into this and will create another PR

**Comment by oscarbenjamin** [comment]
> I will look into this and will create another PR

You don't need to create another PR. If you commit and push further changes then this PR will update.

**Comment by AsRaNi1** [comment]
> > I will look into this and will create another PR
> 
> You don't need to create another PR. If you commit and push further changes then this PR will update.

Yes sorry, I meant I will commit the changes!

**Comment by AsRaNi1** [comment]
@oscarbenjamin what could be the possible problem here?
could you please help me out!

**Comment by oscarbenjamin** [comment]
The tests are failing when running under pypy. Have you tried to run the tests with pypy?

**Comment by AsRaNi1** [comment]
> The tests are failing when running under pypy. Have you tried to run the tests with pypy?

I ran the code quality check with pypy3, and it ran successfully.

**Comment by oscarbenjamin** [comment]
The code quality check is not failing. The failures are in the ODE tests.

**Comment by AsRaNi1** [comment]
> The tests are failing when running under pypy. Have you tried to run the tests with pypy?

On my device too the ODE test is failing, I am unable to ascertain where the origin of the error is. Do the tests somehow tell me where the error is originating from? .i.e from which file?

**Comment by oscargus** [comment]
```
______ sympy/solvers/ode/tests/test_single.py:test_Riccati_special_minus2 ______
Traceback (most recent call last):
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 555, in test_Riccati_special_minus2
    _ode_solver_test(_get_examples_ode_sol_riccati)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 185, in _ode_solver_test
    result = _test_particular_example(example['hint'], example, solver_flag=True)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 235, in _test_particular_example
    dsolve_sol = dsolve(eq, func, simplify=simplify_flag,hint=our_hint)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/ode.py", line 607, in dsolve
    x0=x0, n=n, **kwargs)
  File "/home/runner/work/sympy/sympy/sympy/solvers/deutils.py", line 234, in _desolve
    raise ValueError(string + str(eq) + " does not match hint " + hint)
ValueError: ODE f(x)**2 + 2*Derivative(f(x), x) - f(x)/x + 3/x**2 does not match hint Riccati_special_minus2
________________________________________________________________________________
____________ sympy/solvers/ode/tests/test_single.py:test_Bernoulli _____________
Traceback (most recent call last):
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 564, in test_Bernoulli
    _ode_solver_test(_get_examples_ode_sol_bernoulli)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 185, in _ode_solver_test
    result = _test_particular_example(example['hint'], example, solver_flag=True)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 235, in _test_particular_example
    dsolve_sol = dsolve(eq, func, simplify=simplify_flag,hint=our_hint)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/ode.py", line 607, in dsolve
    x0=x0, n=n, **kwargs)
  File "/home/runner/work/sympy/sympy/sympy/solvers/deutils.py", line 234, in _desolve
    raise ValueError(string + str(eq) + " does not match hint " + hint)
ValueError: ODE x*Derivative(f(x), x) - f(x)**2 + f(x) does not match hint Bernoulli
________________________________________________________________________________
____________ sympy/solvers/ode/tests/test_single.py:test_1st_linear ____________
Traceback (most recent call last):
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 568, in test_1st_linear
    _ode_solver_test(_get_examples_ode_sol_1st_linear)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 185, in _ode_solver_test
    result = _test_particular_example(example['hint'], example, solver_flag=True)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 235, in _test_particular_example
    dsolve_sol = dsolve(eq, func, simplify=simplify_flag,hint=our_hint)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/ode.py", line 607, in dsolve
    x0=x0, n=n, **kwargs)
  File "/home/runner/work/sympy/sympy/sympy/solvers/deutils.py", line 268, in _desolve
    raise ValueError(string + str(eq) + " does not match hint " + hint)
ValueError: ODE -x**2 + x*f(x) + Derivative(f(x), x) does not match hint 1st_linear
________________________________________________________________________________
__________ sympy/solvers/ode/tests/test_single.py:test_almost_linear ___________
Traceback (most recent call last):
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 572, in test_almost_linear
    _ode_solver_test(_get_examples_ode_sol_almost_linear)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 185, in _ode_solver_test
    result = _test_particular_example(example['hint'], example, solver_flag=True)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 235, in _test_particular_example
    dsolve_sol = dsolve(eq, func, simplify=simplify_flag,hint=our_hint)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/ode.py", line 607, in dsolve
    x0=x0, n=n, **kwargs)
  File "/home/runner/work/sympy/sympy/sympy/solvers/deutils.py", line 234, in _desolve
    raise ValueError(string + str(eq) + " does not match hint " + hint)
ValueError: ODE x**2*f(x)**2*Derivative(f(x), x) + f(x)**3 + 1 does not match hint almost_linear
________________________________________________________________________________
 sympy/solvers/ode/tests/test_single.py:test_nth_order_linear_euler_eq_homogeneous 
Traceback (most recent call last):
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 601, in test_nth_order_linear_euler_eq_homogeneous
    _ode_solver_test(_get_examples_ode_sol_euler_homogeneous)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 185, in _ode_solver_test
    result = _test_particular_example(example['hint'], example, solver_flag=True)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 235, in _test_particular_example
    dsolve_sol = dsolve(eq, func, simplify=simplify_flag,hint=our_hint)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/ode.py", line 607, in dsolve
    x0=x0, n=n, **kwargs)
  File "/home/runner/work/sympy/sympy/sympy/solvers/deutils.py", line 268, in _desolve
    raise ValueError(string + str(eq) + " does not match hint " + hint)
ValueError: ODE 2*x**2*Derivative(f(x), (x, 2)) - 3*x*Derivative(f(x), x) does not match hint nth_linear_euler_eq_homogeneous

________________________________________________________________________________
_______ sympy/solvers/ode/tests/test_single.py:test_linear_coefficients ________
Traceback (most recent call last):
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 358, in test_linear_coefficients
    _ode_solver_test(_get_examples_ode_sol_linear_coefficients)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 185, in _ode_solver_test
    result = _test_particular_example(example['hint'], example, solver_flag=True)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 229, in _test_particular_example
    raise AssertionError(message)
AssertionError: Hint did not match the example linear_coeff_01.

The ODE is:
Derivative(f(x), x) + (2*f(x) + 3)/(x + 3).

The expected hint was
linear_coefficients
________________________________________________________________________________
__________ sympy/solvers/ode/tests/test_single.py:test_Airy_equation ___________
Traceback (most recent call last):
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 393, in test_Airy_equation
    _ode_solver_test(_get_examples_ode_sol_2nd_linear_airy)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 185, in _ode_solver_test
    result = _test_particular_example(example['hint'], example, solver_flag=True)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 229, in _test_particular_example
    raise AssertionError(message)
AssertionError: Hint did not match the example 2nd_lin_airy_01.

The ODE is:
-x*f(x) + Derivative(f(x), (x, 2)).

The expected hint was
2nd_linear_airy
________________________________________________________________________________
 sympy/solvers/ode/tests/test_single.py:test_nth_linear_constant_coeff_var_of_parameters 
Traceback (most recent call last):
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 467, in test_nth_linear_constant_coeff_var_of_parameters
    _ode_solver_test(_get_examples_ode_sol_nth_linear_var_of_parameters)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 185, in _ode_solver_test
    result = _test_particular_example(example['hint'], example, solver_flag=True)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 229, in _test_particular_example
    raise AssertionError(message)
AssertionError: Hint did not match the example var_of_parameters_13.

The ODE is:
-2*x - exp(I*x) + Derivative(f(x), x) + 2*Derivative(f(x), (x, 3)) + Derivative(f(x), (x, 5)).

The expected hint was
nth_linear_constant_coeff_variation_of_parameters
________________________________________________________________________________
_______ sympy/solvers/ode/tests/test_single.py:test_nth_order_reducible ________
Traceback (most recent call last):
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 535, in test_nth_order_reducible
    _ode_solver_test(_get_examples_ode_sol_nth_order_reducible)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 185, in _ode_solver_test
    result = _test_particular_example(example['hint'], example, solver_flag=True)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 229, in _test_particular_example
    raise AssertionError(message)
AssertionError: Hint did not match the example reducible_04.

The ODE is:
2*Derivative(f(x), x) + Derivative(f(x), (x, 2)).

The expected hint was
nth_order_reducible
________________________________________________________________________________
 sympy/solvers/ode/tests/test_single.py:test_nth_order_linear_euler_eq_nonhomogeneous_undetermined_coefficients 
Traceback (most recent call last):
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 615, in test_nth_order_linear_euler_eq_nonhomogeneous_undetermined_coefficients
    _ode_solver_test(_get_examples_ode_sol_euler_undetermined_coeff)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 185, in _ode_solver_test
    result = _test_particular_example(example['hint'], example, solver_flag=True)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 229, in _test_particular_example
    raise AssertionError(message)
AssertionError: Hint did not match the example euler_undet_01.

The ODE is:
Eq(x**2*Derivative(f(x), (x, 2)) + x*Derivative(f(x), x), 1).

The expected hint was
nth_linear_euler_eq_nonhomogeneous_undetermined_coefficients
```

**Comment by oscarbenjamin** [comment]
The cause of the failures must be one of the removed `__ne__` methods. First make sure that you can quickly reproduce the test failure (locally). Then I suggest to try adding back the `__ne__` methods one by one until the tests pass. That way you can identify which `__ne__` method affects these tests.

**Pull request closed without merging**